Beam/Include/Beam/Threading/Sync.hpp

```diff
@@ -52,38 +130,53 @@ namespace Threading {
       */
       Sync(Sync&& sync);
 
-      //! Returns a reference of the value.
-      T& AcquireReference();
-
       //! Returns a copy of the value.
       T Acquire() const;
 
+      //! Acquire Sync's value in a synchronized manner.
+      /*!
+        \param f The action to perform on the value.
+      */
       template<typename F>
-      decltype(auto) With(F f) {
-        boost::unique_lock<Mutex> lock(m_mutex);
-        m_lock = &lock;
-        return f(m_value);
+      decltype(auto) With(const F& f) {
+        if constexpr(boost::hof::is_invocable<F, Value>()) {
```

Comment:
Had to use `boost::hof` as `std::is_invocable` is bugged in VS. The latter will return `true` even if attempting to bind non-reference type to a callable taking a non-const reference; it correctly produces `false` in GCC and CLang though.